app/models/fact_value.rb

```diff
@@ -81,13 +81,15 @@ def self.count_each(fact, options = {})
     output
   end
 
-  def self.build_facts_hash(facts)
-    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
-
+  def self.build_facts_hash(facts, has_host = false)
     hash = {}
     facts.each do |fact|
+      hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
```

Comment:
This line should be outside of `each`, otherwise it will be executed once per each fact.